### PR TITLE
feature: S3C-4486 Add metric for replication status

### PR DESCRIPTION
**Pull request by miniscruff**
* generalized metric utils from queue populator

Will resolve conflicts after reviews

### CODE REVIEW

File: tests/unit/lib/util/metrics.spec.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+const sinon = require('sinon');
+const { wrapCounterInc, wrapGaugeSet } =
+    require('../../../../lib/util/metrics');
+
+describe.only('Metrics', () => {
```
Comment (naren-rajendran) [inline tests/unit/lib/util/metrics.spec.js]: only?

Comment (miniscruff) [reply tests/unit/lib/util/metrics.spec.js]: grrr

**Comment by miniscruff** [comment]
ping

File: extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js [hunk 2]
```diff
@@ -16,13 +16,43 @@ const ObjectQueueEntry = require('../utils/ObjectQueueEntry');
...
+
+const replicationStatusMetric = new promClient.Counter({
+    name: 'replication_status_changed_total',
+    help: 'Number of objects updated',
+    labelNames: ['origin', 'containerName', 'replicationStatus'],
+});
+
+const kafkaLagMetric = new promClient.Gauge({
+    name: 'kafka_lag',
+    help: 'Number of objects behind we are in the Kafka log',
```
Comment (jonathan-gramain) [inline extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js]:
```suggestion
    help: 'Number of update entries waiting to be consumed from the Kafka topic',
```

File: extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js [hunk 3]
```diff
@@ -16,13 +16,43 @@ const ObjectQueueEntry = require('../utils/ObjectQueueEntry');
...
+    labelNames: ['origin', 'containerName', 'partition'],
+});
+
+/**
+ * Contains methods to incrememt different metrics
+ * @typedef {Object} MetricsHandler
+ * @property {CounterInc} status - Increments the replication status metric
```
Comment (jonathan-gramain) [inline extensions/replication/replicationStatusProcessor/ReplicationStatusProcessor.js:50]: You might want to document `lag` as well

**Review by jonathan-gramain — APPROVED** [review]

**Comment by naren-rajendran** [comment]
Are operations, bandwidth and latency handled separately (a separate pr)? or should they be extrapolated with the current changes? otherwise the commit looks good to me!

**Comment by miniscruff** [comment]
> Are operations, bandwidth and latency handled separately (a separate pr)? or should they be extrapolated with the current changes? otherwise the commit looks good to me!

Not sure what operations are referring to here, like objects updated or Kafka messages? What is an "operation"?
Not sure what you want to measure for bandwidth, like the total size of all metadata update network packets? For status processor, I am not sure this is required, for queue processor that pulls and pushes objects I can see it tracking that.
Not entirely sure what latency is referring to here, network latency between the status processor and kafka? Time to update one object status? Just saying latency is not enough for me.

Note that none of these are part of this PR yes, I had not planned on anything more than what I created here. The ticket was specifically targeting "Add prometheus metrics for Update Metadata in Source graph on Status Processor dashboard"

**Comment by naren-rajendran** [comment]
> > Are operations, bandwidth and latency handled separately (a separate pr)? or should they be extrapolated with the current changes? otherwise the commit looks good to me!
> 
> Not sure what operations are referring to here, like objects updated or Kafka messages? What is an "operation"?
> Not sure what you want to measure for bandwidth, like the total size of all metadata update network packets? For status processor, I am not sure this is required, for queue processor that pulls and pushes objects I can see it tracking that.
> Not entirely sure what latency is referring to here, network latency between the status processor and kafka? Time to update one object status? Just saying latency is not enough for me.
> 
> Note that none of these are part of this PR yes, I had not planned on anything more than what I created here. The ticket was specifically targeting "Add prometheus metrics for Update Metadata in Source graph on Status Processor dashboard"

These (bandwidth, operations & latency) are from the requirements and the mock charts, https://scality.github.io/citadel/requirements/s3-replication-monitoring/ I couldn't find any s3c tickets related to these metrics and I was wondering if they are going to be handled.

**Review by naren-rajendran — APPROVED** [review]

**Pull request merged**